metric_providers/base.py

```diff
@@ -47,8 +47,17 @@ def __init__(
         if not self._skip_check:
             self.check_system()
 
-    # this is the default function that will be overridden in the children
+    # this is the default function that can be overridden in the children
+    # by default we expect the executable to have a -c switch to test functionality
     def check_system(self):
+        call_string = self._metric_provider_executable
+        if self._metric_provider_executable[0] != '/':
+            call_string = f"{self._current_dir}/{call_string}"
+
+        ps = subprocess.run([f"{call_string}", '-c'], capture_output=True, encoding='UTF-8', check=False)
+        if ps.returncode != 0:
+            raise MetricProviderConfigurationError(f"{self._metric_name} provider could not be started.\nError: {ps.stderr}\nAre you running in a VM / cloud / shared hosting?\nIf so please disable the {self._metric_name} provider in the config.yml")
+
```

Comment:
This implementation doesn't quite work, as the call_string is not always the binary. For NetworkConnectionsProxyContainerProvider for example, the   callstring is "/usr/bin/tinyproxy".

Additionally, wouldn't this require all providers with a binary to include this -c flag? Is that something we want to enforce moving forward?

I suggest we move this functionality in its own function (CheckMetricProviderBinary) , and explictely call it in whichever providers its applicable to (which admittedly is most of them now)

Reply:
For now, I fixed the NetworkConnectionsProxyContainerProvider by moving the simulatenous provider  check into its own function (that's still called by the base check_system function), and changed that specific provider to call the new function instead of super.check_system().

Reply:
I am not a fan of the additional function that technically belongs to the check_system logically.

The reason for having this function seems that it can be overriden with a different call to the binary for checking.

If that is the reason, why don't we make rather a variable that specifies how the default call is as a python list ala `binary_check=["metric_provider_binary", "-c"]`

This way it is way cleaner and there is no risk of not calling the functions correctly. Flow should be that you cann the super.check_system() and then add logic on top, or override the function alltogether.
Having a mix of not calling the super function, adding own logic and then calling another super / object method (currently check_parallel...) seems hard to follow

Reply:
I unresolved this as the question is not answered and I saw no commit on chaning this yet

Reply:
For the record on this - I had marked this as resolved after my last comment, before you had asked any additional questions - I thought it was ok as I was the one who opened this issue, answered my own question, and then resolved it.

To answer your question - first, the check_parallel_provider is part of and called by the check_system. It was split out for the cases where we want to do a different kind of system check, but still also want to check parallel providers.

I did think of adding adding parameters for the check_system with what the binary call would be - but this doesn't cover all of our bases. I can think of at least one provider where the check requires overwriting this function, but we still want to check the parallel providers. 

I see that the flow of overriding the function, but still calling another part of the function can be confusing though. What we could do instead is remove it from check_system, but instead earlier in the base.py do this:

```
        if not self._skip_check:
            self.check_system()
            self.check_parallel_provider_running()
```

that way a metric provider that wants to have a different kind of system check, but still keep the parallel provider check wouldn't have to remember to call the super().check_parallel_provider_running(). I don't see this as really resolving anything though, as then the providers which *don't* need to run the check_parallel_provider_running (xgboost/sdia, as you pointed out before), then need to remember to pass these as well, so its just moving the mental load over to a different place. Though - in the case its forgotten, its better to accidentally have an extra check than accidentally have one less check.